tinygrad/runtime/ops_qcom.py

```diff
@@ -0,0 +1,269 @@
+    self.cmd(adreno.CP_SET_MARKER, adreno.RM6_COMPUTE)
+    self.reg(adreno.REG_A6XX_HLSQ_CONTROL_2_REG, 0xfcfcfcfc, 0xfcfcfcfc, 0xfcfcfcfc, 0xfc, 0x140)
+    self.reg(adreno.REG_A6XX_HLSQ_INVALIDATE_CMD, 0x60)
+    self.reg(adreno.REG_A6XX_HLSQ_INVALIDATE_CMD, 0x0)
+    self.reg(adreno.REG_A6XX_SP_CS_TEX_COUNT, 0x80)
+    self.reg(adreno.REG_A6XX_SP_CS_IBO_COUNT, 0x40)
+    self.reg(adreno.REG_A6XX_SP_MODE_CONTROL, adreno.ISAMMODE_CL << adreno.A6XX_SP_MODE_CONTROL_ISAMMODE__SHIFT)
+    self.reg(adreno.REG_A6XX_SP_PERFCTR_ENABLE, adreno.A6XX_SP_PERFCTR_ENABLE_CS)
+    self.reg(adreno.REG_A6XX_SP_TP_MODE_CNTL, adreno.ISAMMODE_CL | (1 << 3)) # ISAMMODE|UNK3
+    self.reg(adreno.REG_A6XX_TPL1_DBG_ECO_CNTL, 0)
+    self.reg(
+      adreno.REG_A6XX_HLSQ_CS_NDRANGE_0,
+        # kernel dimenstion = 3
+        3 | ((local_size[0] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEX__SHIFT)
+        | ((local_size[1] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEY__SHIFT)
+        | ((local_size[2] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEZ__SHIFT),
+        global_size_mp[0], 0, global_size_mp[1], 0, global_size_mp[2], 0, # global size x,y,z followed by offsets
+        0xccc0cf, 0x2fc,
+        global_size[0], global_size[1], global_size[2], # original global sizes
+    )
+    self.reg(adreno.REG_A6XX_SP_CHICKEN_BITS, 0x20)
+    self.reg(
+      adreno.REG_A6XX_SP_CS_CTRL_REG0,
+      (adreno.THREAD128 << adreno.A6XX_SP_CS_CTRL_REG0_THREADSIZE__SHIFT) 
+      | (prg.halfreg << adreno.A6XX_SP_CS_CTRL_REG0_HALFREGFOOTPRINT__SHIFT)
+      | (prg.fullreg << adreno.A6XX_SP_CS_CTRL_REG0_FULLREGFOOTPRINT__SHIFT)
+      | (16 << adreno.A6XX_SP_CS_CTRL_REG0_BRANCHSTACK__SHIFT),
+      0x41, 0, prg.prg_offset, # offsets
+      *data64_le(prg.lib_gpu.va_addr), 0, *data64_le(prg.private_gpu.va_addr), prg.private_gpu.size,
+    )
+    self.reg(adreno.REG_A6XX_SP_CS_CONFIG, 0x100)
+    self.reg(adreno.REG_A6XX_SP_CS_PVT_MEM_HW_STACK_OFFSET, 0)
+    self.cmd(adreno.CP_LOAD_STATE6_FRAG,
+             (adreno.ST_CONSTANTS << adreno.CP_LOAD_STATE6_0_STATE_TYPE__SHIFT)
+             | (adreno.SS6_INDIRECT << adreno.CP_LOAD_STATE6_0_STATE_SRC__SHIFT)
+             | (adreno.SB6_CS_SHADER << adreno.CP_LOAD_STATE6_0_STATE_BLOCK__SHIFT)
+             | (256 << adreno.CP_LOAD_STATE6_0_NUM_UNIT__SHIFT),
+             *data64_le(kernargs))
+    self.cmd(adreno.CP_LOAD_STATE6_FRAG,
+            (adreno.ST_SHADER << adreno.CP_LOAD_STATE6_0_STATE_TYPE__SHIFT)
+            | (adreno.SS6_INDIRECT << adreno.CP_LOAD_STATE6_0_STATE_SRC__SHIFT)
+            | (adreno.SB6_CS_SHADER << adreno.CP_LOAD_STATE6_0_STATE_BLOCK__SHIFT)
+            | (math.ceil(prg.image_size / 128) << adreno.CP_LOAD_STATE6_0_NUM_UNIT__SHIFT),
+            *data64_le(prg.lib_gpu.va_addr))
+    self.reg(adreno.REG_A6XX_SP_CS_INSTRLEN, prg.image_size // 4)
+    self.cmd(adreno.CP_RUN_OPENCL, 0)
+    self.cmd(adreno.CP_WAIT_FOR_IDLE)
+
+class QcomCopyQueue(HWCopyQueue, QcomComputeQueue):
```

Comment:
copy queue will be optional in master soon

Reply:
Nice! Done